ucte/ucte-converter/src/main/java/com/powsybl/ucte/converter/CounterNamingStrategy.java

```diff
@@ -0,0 +1,232 @@
+            char letter = (char) ('A' + ((namingCounter - BASE_LETTER_OFFSET) / MODULO_VALUE));
+            return String.format("%04d%c", baseNumber, letter);
+        }
+        return String.format("%05d", namingCounter);
+    }
+
+    private static char getUcteVoltageLevelCode(double voltage) {
+        if (voltage < UcteVoltageLevelCode.VL_27.getVoltageLevel()) {
+            return DEFAULT_VOLTAGE_CODE;
+        }
+        if (voltage > UcteVoltageLevelCode.VL_750.getVoltageLevel()) {
+            return '0';
+        }
+
+        return Arrays.stream(UcteVoltageLevelCode.values())
+                .min(Comparator.comparingDouble(code ->
+                        Math.abs(voltage - code.getVoltageLevel())))
+                .map(code -> (char) ('0' + code.ordinal()))
+                .orElse(DEFAULT_VOLTAGE_CODE);
+    }
+
+    private UcteCountryCode getCountryCode(VoltageLevel voltageLevel) {
+        Country country = voltageLevel.getSubstation()
+                .flatMap(Substation::getCountry)
+                .orElseThrow(() -> new UcteException(NO_COUNTRY_ERROR));
+
+        try {
+            return UcteCountryCode.valueOf(country.name());
+        } catch (IllegalArgumentException e) {
+            throw new UcteException(String.format(NO_UCTE_COUNTRY_ERROR, country.getName()));
+        }
+    }
+
+    private UcteElementId generateUcteElementId(String id, UcteNodeCode node1, UcteNodeCode node2) {
+        if (ucteElementIds.containsKey(id)) {
+            return ucteElementIds.get(id);
+        }
+
+        return ORDER_CODES.stream()
+                .map(orderCode -> new UcteElementId(node1, node2, orderCode))
+                .filter(elementId -> !ucteElementIds.containsValue(elementId))
+                .findFirst()
+                .map(elementId -> {
+                    ucteElementIds.put(id, elementId);
+                    return elementId;
+                })
+                .orElseThrow(() -> new UcteException("Unable to generate unique element ID"));
+    }
+
+    private UcteElementId generateUcteElementId(Branch<?> branch) {
+        if (ucteElementIds.containsKey(branch.getId())) {
+            return ucteElementIds.get(branch.getId());
+        }
+        UcteNodeCode node1 = ucteNodeIds.get(branch.getTerminal1().getBusBreakerView().getBus().getId());
+        UcteNodeCode node2 = ucteNodeIds.get(branch.getTerminal2().getBusBreakerView().getBus().getId());
+        return generateUcteElementId(branch.getId(), node1, node2);
+    }
+
+    private UcteElementId generateUcteElementId(DanglingLine danglingLine) {
+        if (ucteElementIds.containsKey(danglingLine.getId())) {
+            return ucteElementIds.get(danglingLine.getId());
+        }
+
+        Bus bus = danglingLine.getTerminal().getBusBreakerView().getBus();
+        VoltageLevel voltageLevel = danglingLine.getTerminal().getVoltageLevel();
+
+        generateUcteNodeId(danglingLine.getPairingKey(), voltageLevel);
```

Comment:
pairingKey could be null, you should:
 - first try we pairing key if not null
 - otherwise use the dangling line ID